File: frontend/src/components/projects/binomial/calc/binomial.ts
import { OptionType, BinomialLatticeType } from "../../util/common_types.types";

const calculateBinomial = (
  S: number,
  K: number,
  T: number,
  r: number,
  sigma: number,
  steps: number,
  optionType: OptionType,
  latticeType: BinomialLatticeType = "CRR"
): number => {
  const dt = T / steps;

  const {u, d, p} = getLatticeParameters(dt, r, sigma, latticeType);
  const discountFactor = Math.exp(-r * dt);

  const assetPrices = new Float64Array(steps + 1);
  const optionValues = new Float64Array(steps + 1);

  for (let i = 0; i <= steps; i++) {
    assetPrices[i] = S * Math.pow(u, steps - i) * Math.pow(d, i);
  }

  for (let i = 0; i <= steps; i++) {
    optionValues[i] =
      optionType === "call"
        ? Math.max(0, assetPrices[i] - K)
        : Math.max(0, K - assetPrices[i]);
  }

  for (let step = steps - 1; step >= 0; step--) {
    for (let i = 0; i <= step; i++) {
      optionValues[i] =
        discountFactor * (p * optionValues[i] + (1 - p) * optionValues[i + 1]);
    }
  }

  return optionValues[0];
};

const getLatticeParameters = (
  dt: number,
  r: number,
  sigma: number,
  latticeType: BinomialLatticeType
): { u: number; d: number; p: number } => {
  let u: number, d: number, p: number;
  switch (latticeType) {
    case "CRR":
      u = Math.exp(sigma * Math.sqrt(dt));
      d = 1 / u;
      p = (Math.exp(r * dt) - d) / (u - d);
      break;
    case "JR":
      u = Math.exp((r - 0.5 * sigma ** 2) * dt + (sigma * Math.sqrt(dt)));
      d = Math.exp((r - 0.5 * sigma ** 2) * dt - (sigma * Math.sqrt(dt)));
      p = 0.5;
      break;
    case "TIAN":
      const a = Math.exp(r * dt);
      const b = Math.exp(sigma * Math.sqrt(dt));
      u = 0.5 * a * b * (b + 1 + Math.sqrt(b * b + 2 * b - 3));
      d = 0.5 * a * b * (b + 1 - Math.sqrt(b * b + 2 * b - 3));
      p = (a - d) / (u - d);
      break;  
    case "TRG":
      u = Math.exp(sigma * Math.sqrt(dt) + (r - 0.5 * sigma * sigma) * dt);
      d = Math.exp(-sigma * Math.sqrt(dt) + (r - 0.5 * sigma * sigma) * dt);
      p = 0.5;
      break;
    default:
      u = Math.exp(sigma * Math.sqrt(dt));
      d = 1 / u;
      p = (Math.exp(r * dt) - d) / (u - d);
      break;
  }
  return { u, d, p };
}

export default calculateBinomial;

File: backend/app/services/binomial.py
import numpy as np
import math


def simulate_binomial(
  S, K, T, r, sigma, q, steps, option_type, option_style, lattice_type 
):
  # TODO : NOW MAKE IT WORK FOR DIFFERENT OPTION STYLES (AMERICAN, EUROPEAN, ASIAN, BARRIER)
  dt = T/steps

  u, d, p = get_lattice_parameters(dt, r, sigma, q, lattice_type)

  discount_factor = np.exp(-(r+q) * dt)

  C = S * d ** (np.arange(steps, -1, -1)) * u ** (np.arange(0, steps+1, 1))
  if option_type == "call":
    C = np.maximum(np.zeros(steps+1), C-K)
  elif option_type == "put":
    C = np.maximum(np.zeros(steps+1), K-C)
  else:
    raise ValueError("Value should only be a call or a put.")

  for i in range(steps, 0, -1):
    if option_style == "european":
      C = discount_factor * ( (C[1:i+1] * p) + (C[0:i] * (1-p)) )
    elif option_style == "american":
      S_prices = S * (d ** np.arange(i-1, -1, -1)) * (u ** np.arange(0, i))
      new_C = discount_factor * (p * C[1:i+1] + (1 - p) * C[0:i])
      C = new_C 
      if option_type == "put":
        C = np.maximum(C, K - S_prices)
      else:
        C = np.maximum(C, S_prices - K)

  binomial_price = C[0]

  return { "binomial_price" : binomial_price }
  

def get_lattice_parameters(dt, r, sigma, q, lattice_type):
  if lattice_type == "CRR":
    u = math.exp(sigma * math.sqrt(dt))
    d = 1 / u
    p = (math.exp((r-q) * dt) - d) / (u - d)

  elif lattice_type == "JR":
    u = math.exp(((r-q) - 0.5 * sigma**2) * dt + sigma * math.sqrt(dt))
    d = math.exp(((r-q) - 0.5 * sigma**2) * dt - sigma * math.sqrt(dt))
    p = 0.5

  elif lattice_type == "TIAN":
    a = math.exp((r-q) * dt)
    b = math.exp((sigma**2) * dt)
    u = 0.5 * a * b * (b + 1 + math.sqrt(b**2 + 2*b - 3))
    d = 0.5 * a * b * (b + 1 - math.sqrt(b**2 + 2*b - 3))
    p = (a - d) / (u - d)

  elif lattice_type == "TRG":
    u = math.exp(sigma * math.sqrt(dt) + ((r-q) - 0.5 * sigma*sigma) * dt)
    d = math.exp(-sigma * math.sqrt(dt) + ((r-q) - 0.5 * sigma*sigma) * dt)
    p = 0.5

  else:
    u = math.exp(sigma * math.sqrt(dt))
    d = 1 / u
    p = (math.exp((r-q) * dt) - d) / (u - d)

  return u, d, p

